src/common/io/socket/common.c

```diff
@@ -126,3 +128,86 @@ sckOptionSet(int fd)
 
     FUNCTION_TEST_RETURN_VOID();
 }
+
+/***********************************************************************************************************************************
+Use poll() to determine when data is ready to read/write on a socket. Retry after EINTR with whatever time is left on the timer.
+***********************************************************************************************************************************/
+// Helper to determine when poll() should be retried
+static bool
+sckReadyRetry(int pollResult, int errNo, bool first, Wait *wait)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(INT, pollResult);
+        FUNCTION_TEST_PARAM(INT, errNo);
+        FUNCTION_TEST_PARAM(BOOL, first);
+        FUNCTION_TEST_PARAM(WAIT, wait);
+    FUNCTION_TEST_END();
+
+    // No retry by default
+    bool result = false;
+
+    // Process errors
+    if (pollResult == -1)
+    {
+        // Don't error on an interrupt. If the interrupt lasts long enough there may be a timeout, though.
+        if (errNo != EINTR)
+            THROW_SYS_ERROR_CODE(errNo, KernelError, "unable to poll socket");
+
+        // Retry if first iteration or time remaining
+        result = first || waitMore(wait);
```

Comment:
Hmm, yeah, good point. That also gets rid of `first` since we don't need to skip at wait.

On top of that waitNew() allocates memory which is not ideal here. I had just realized this code needed to be wrapped in a mem context but this will avoid that. Wait was just bad for this application.

Reply:
I couldn't get rid of first without using some other param in a weird way, but I did the rest.